[*chan] Add text-posts opt (generic text-tweets)

wlritchi opened the pull request:
The Twitter extractor has a `text-tweets` option, which causes it to return Directory results even for tweets with no media. Some other extractors (such as Patreon) behave this way by default. Still others skip non-media posts entirely. This PR adds a `text-posts` option to (some of?) the extractors that otherwise skip non-media posts, and updates the Twitter extractor to respect this new option alongside the existing Twitter-specific one.

I did a cursory search for common patterns (`if not files` and the like) that I expected might appear in extractors that skipped non-media posts. The only ones I found were the *chan extractors. The logic change is broadly similar in all of them: move the Directory result inside the loop over posts rather than just one per thread.

This change probably isn't meaningful without corresponding post-processor config for each extractor. I tested with the following config additions:

```
...
    "postprocessors": [
        {
            "name": "metadata",
            "event": "post",
            "filename": "{no}.post_data.json",
            "whitelist": ["2chan", "4chan", "420chan", "vichan"]
        },
        {
            "name": "metadata",
            "event": "post",
            "filename": "{postId}.post_data.json",
            "whitelist": ["8chan", "lynxchan"]
        }
    ]
```

420chan appears to be permanently offline, so no test results for that one. For the rest, I found threads with some text-only posts[^ew] and manually tested. Everything seems to work as I expected.

This change was requested in https://github.com/mikf/gallery-dl/issues/570#issuecomment-850513297. The ticket itself appears to be about Twitter; should that one be closed as completed? (regardless of this PR)

[^ew]: Which was a rather unpleasant task for some of these sites.

wlritchi committed:
[*chan] Add text-posts opt (generic text-tweets)

wlritchi commented:
Mmm, further testing suggests that actually the dict update changes in lynxchan and 8chan extractors aren't right; it looks like the threadId is overwriting the postId. Will investigate further.

wlritchi commented:
Oh, and if #3447 is merged, that extractor will need a similar change to the ones here.

mikf committed:
[jpgfish] update domain to 'jpg1.su' (#4494)

mikf committed:
[gelbooru_v02] support 'xbooru.com' (#4493)

mikf committed:
[gfycat] remove module

mikf committed:
[twitter] fix crash when 'sortIndex' is None (#4499)

mikf committed:
[subscribestar] fix preview detection (#4468)

and show a warning message when posts contain previews

mikf committed:
[formatter] use value of last alternative (#4492)

fixes {fieldname|''} evaluating to the value of 'keywords-default'
instead of an empty string

mikf committed:
update test comment positions

always put them above the test they're referring to

mikf committed:
publish export_tests.py script

mikf committed:
add -t and -c command-line arguments to export_tests.py

mikf committed:
fix for categories containing '.'

Files with ':' in their name cannot be imported, as __import__()
will try to interpret them as submodules.

mikf committed:
fix '"' escapes

mikf committed:
add exported extractor results

mikf committed:
remove test results in extractor modules

and add generic example URLs

mikf committed:
update test/test_extractor.py

mikf committed:
update/fix --list-extractors

mikf committed:
update scripts/supportedsites.py

mikf committed:
[redgifs] provide 'collection' metadata in a separate field (#4508)

instead of overwriting the actual metadata

mikf committed:
[behance] handle videos without 'renditions' (#4523)

mikf committed:
[reddit] improve comment metadata (#4482)

- provide 'date'
- make metadata of the main submission available as 'submission[…]'

mikf committed:
update test/test_results.py

mikf committed:
Merge branch 'tests'

mikf committed:
remove another '*' for keyword-only arguments

076380e0

mikf committed:
[recursive] simplify

mikf committed:
[mastodon] support '/@USER/following' URLs

Previously, only '/users/USER/following' got matched.

mikf committed:
fix bug in test_init and update example URLs

mikf committed:
[kemonoparty] update favorites API endpoint (#4522)

mikf committed:
exclude export_tests.py from linting for Python<3.8

Scripter17 commented:
This seems to not work with `--write-metadata`

mikf committed:
do not restrict flake8 version

mikf commented:
`--write-metadata` only write per-file metadata, while the metadata returned here is per-post. You'd need an `"event": "post"` metadata post processor to make use of it.

Scripter17 commented:
Sorry for not being clear

Adding `--write-metadata` stops the post-processor from working. Which I assume is not the intended behaviour

mikf committed:
[twitter] fix '_extractor' of following results (#4536)

regression from 20ed647f

mikf committed:
[architizer] fix extraction (#4537)

mikf commented:
`--write-metadata` will overwrite post-processors defined at the top level of your config, but anything in `extractor` or a site category should be unaffected.

``` json
{
    "postprocessors": [{"name": "mtime"}],
    "extractor": {
        "postprocessors": [{"name": "ugoira"}],
        "4chan": {
            "postprocessors": [{"name": "compare"}]
        }
    }
}
```

```
$ gallery-dl -c config.json -v https://boards.4chan.org/tg/thread/15396072/
[4chan][debug] Active postprocessor modules: [ComparePP, UgoiraPP, MtimePP]

$ gallery-dl -c config.json -v --write-metadata https://boards.4chan.org/tg/thread/15396072/
[4chan][debug] Active postprocessor modules: [ComparePP, UgoiraPP, MetadataPP]
```

Scripter17 commented:
1. That should probably be changed/raise a warning
2. After I got it working, it seems lynxchan posts are all put in different folders instead of one per-thread folder. This happens with and without the postprocessor and `--write-metadata`

mikf committed:
send Referer headers by default

mikf committed:
[mangakakalot] update domain

the old one still works, but it incurs a redirect

mikf committed:
[patreon] fix extraction (#4547)

mikf committed:
[reddit] improve comment metadata v2 (#4482)

provide main submission metadata at the top level
and comment metadata inside the 'comment' field,
i.e. the other way round than in 1710f1e9

mikf committed:
[instagram] handle exceptions due to missing media (#4555)

mikf committed:
fix --range causing crashes (#4557)

regression caused by a383eca7

mikf committed:
[deviantart] update Eclipse API endpoints (#4553)

mikf committed:
[deviantart] disable JWT updates (#4548, #4563)

back to lowres images ...

mikf committed:
[deviantart] add 'is_original' metadata field (#4559)

true for 'downloadable' content, journals, flash animations,
and images without '/v1/' in their URL; false otherwise

mikf committed:
[redgifs] match gfycat image URLs (#4558)

mikf committed:
[reddit] add 'previews' option (#4322)

another way to disable this new behavior

mikf committed:
[danbooru] support 'donmai.moe' URLs

mikf committed:
[pillowfort] extract 'b2_lg_url' media (#4570)

mikf committed:
[pillowfort] support '/tagged/' URLs (#4570)

mikf committed:
[reddit] ignore '/message/compose' URLs (#4482)

mikf committed:
[tests] support 'range()' for #count and metadata checks

mikf committed:
[tests] allow checking for exact URL results

mikf committed:
[deviantart] fix JWT replacement (#293, #4548, #4563)

And again, a huge thank you to @Ironchest337
for discovering this.

mikf committed:
[deviantart] re-add 'quality' option and 'intermediary' transform

mikf committed:
[deviantart] use private tokens for 'is_mature' posts (#4563)

hrxn committed:
[reddit] ignore '/message/compose' URLs without www subdomain

hrxn committed:
Update gallery_dl/extractor/reddit.py

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
merge #4581: [reddit] ignore /message/compose URLs without www subdomain

mikf committed:
fix imports when using the gallery_dl directory as argument (#4581)

mikf committed:
[regifs] support 'order' parameter for user URLs (#4583)

mikf committed:
[tests] support one regex per URL for #pattern

mikf committed:
fix symlink resolution in __main__.py

adapt ytdl order

mikf committed:
[bunkr] fix extraction (#4514, #4532, #4529, #4540)

mikf committed:
[tests] use fallback URLs for content tests (#3163)

mikf committed:
[tests] re-implement filtering by basecategory

mikf committed:
[tests] add 'msg' arguments to assert statements

mikf committed:
[nsfwalbum] disable Referer headers by default (#4598)

mikf committed:
[nsfwalbum] detect '/error.jpg' images (#4598)

mikf committed:
[instagram] better error message for invalid users (#4606)

mikf committed:
[shimmie2] remove 'meme.museum'

mikf committed:
[deviantart] update Eclipse API endpoints 2 (#4615)

mikf committed:
release version 1.26.0

Commit:
[twitter] add possible_sensitive field

mikf committed:
[twitter] fix crash due to missing 'source' (#4620)

regression caused by 06aaedde

hrxn committed:
[imgbb] Fix `user` extraction, add `displayname`

mikf committed:
[patreon] update 'campaign_id' path (#4639)

mikf committed:
[newgrounds] support 'art-image' files (#4642)

mikf committed:
[hentaifoundry] fix '.swf' file downloads (#4641)

mikf committed:
[pp:exec] support more replacement fields for '--exec' (#4633)

- {_directory}
- {_filename}
- {_path} (alias for {})

mikf committed:
[imgbb] fix flake8 and username order

mikf committed:
merge #4626: [imgbb] fix 'user' extraction, add 'displayname'

mikf committed:
[imgbb] update pagination end condition (#4626)

mikf committed:
[twitter] rename to 'sensitive'; use 'tget()'

mikf committed:
merge #4619: [twitter] add 'sensitive' metadata field

mikf committed:
[reddit] fix wrong previews (#4649)

caused by a failed comment URL
using the main submission's preview as fallback

14af15bd
4963bb9b

mikf committed:
[cookies] open cookie databases in read-only mode

bypasses the need to copy the entire database file
might solve #4195

mikf committed:
[moebooru] add 'metadata' option (#4646)

for extended 'pool' metadata

mikf committed:
[deviantart] disable 'jwt' (#4652)

mikf committed:
[deviantart] add a couple 'deactivated account' test URLs

mikf committed:
[deviantart] implement '"group": "skip"' (#4630)

mikf committed:
[newgrounds] improve 'art-image' extraction (#4642)

- download files in original resolution
- replace .webp with extension of first file

mikf committed:
[fantia] add 'content_count' and 'content_num' metadata fields (#4627)

mikf committed:
[warosu] fix extraction (#4634)

mikf committed:
[fantia] bad workaround for 833dce14 (#4627)

at least this makes "filter": "content_num == content_count+1"
with "event": "post-after" work

mikf committed:
use Python 3.12 for GitHub Actions

mikf committed:
[imgbb] update username extraction (#4626)

mikf committed:
[kemonoparty] support discord URLs with channel IDs (#4662)

mikf committed:
[kemonoparty] fix DM dates

mikf committed:
[chevereto] add generic extractors (#4664)

- support jpgfish
- support pixl.li / pixl.is (#3179, #4357)

enduser420 committed:
[redgifs] fix 'niches' extraction

mikf committed:
merge #4667: [redgifs] fix 'niches' extraction (#4666)

mikf committed:
[chevereto] support 'img.kiwi' and 'deltaporno.com' (#4664, #1381)

mikf committed:
[tests] fix spurious failures in '_assert_isotime()'

mikf committed:
[deviantart] expand nested comment replies (#4653)

mikf committed:
[tests] implement 'len:'

sanddudu committed:
[kemonoparty] update API endpoint (#4676)

sanddudu committed:
fix line too long

mikf committed:
merge #4676: [kemonoparty][coomerparty] update API endpoints (#4676)

mikf committed:
[kemonoparty] various updates and fixes (#4676, #4681)

- fix pagination
- fix 'date' metadata
- fix discord channel API endpoint

mikf committed:
[kemonoparty] restore discord pagination (#4676)

mikf committed:
[kemonoparty] initial support for post revisions (#4498, #4597)

- single revision
https://kemono.party/SERVICE/user/12345/post/12345/revision/12345

- all revisions
https://kemono.party/SERVICE/user/12345/post/12345/revisions

mikf committed:
[kemonoparty] support post searches (#3385, #4057)

mikf committed:
[kemonoparty] add 'revisions' option (#4498, #4597)

mikf committed:
[bunkr] add extractor for media URLs (#4684)

mikf committed:
[bunkr] fix '/d/' file URLs (#4685)

mikf committed:
[4chanarchives] disable Referer headers by default (#4686)

mikf committed:
[sankaku] support '/posts/' URLs (#4688)

mikf committed:
[newgrounds] support 'imageData' files (#4642)

mikf committed:
[cookies] include exception in fallback warning

mikf committed:
release version 1.26.1

interruptinuse committed:
[reddit] implement Reddit Mobile share links

enduser420 committed:
[4archive] add 'thread' and 'board' extractors

mikf committed:
[twitter] cache 'user_by_…' results (#4719)

mikf committed:
[patreon] improve 'campaign_id' handling (#4699, #4715)

- add ways to directly specify a 'campaign_id'
  - 'campaign-id' config option
  - 'c' or 'campaign_id' URL query parameter
- more descriptive error messages
- show 'campaign_id' value in debug log

mikf committed:
[misskey] support 'misskey.design' (#4713)

mikf committed:
merge #4693: [reddit] support Reddit Mobile share links

enduser420 committed:
[4archive] update

mikf committed:
[4archive] docs/supportedsites

mikf committed:
merge #4714: [4archive] add 'thread' and 'board' extractors

(#1262, #2418, #4400, #4710)

mikf committed:
[idolcomplex] disable Referer headers by default (#4726)

mikf committed:
[kemonoparty] fix parsing of non-standard 'dates' (#4676)

mikf committed:
[kemonoparty] add 'revision_index' metadata field (#4727)

mikf committed:
[instagram] update API headers

mikf committed:
[twitter] fix avatars without 'date' information (#4696)

mikf committed:
[twitter] restore truncated retweet text (#3430, #4690)

Tobi823 committed:
Try to parse newer HTTP response from Patreon

mikf committed:
[sankaku] support '/posts/' tag search URLs (#4740)

mikf committed:
prevent crash when 'stdout.line_buffering' is not defined (#642)

mikf committed:
[patreon] refactor _extract_bootstrap()

mikf committed:
merge #4738: [patreon] parse new bootstrap data format (#4736)

mikf committed:
[exhentai] fix extraction (#4730)

- update to new API response layout
- use proper API server URL
- fix 'filesize' metadata

mikf committed:
[behance] unescape embed URLs (#4742)

mikf committed:
[weibo] fix Sina Visitor request

Commit:
[twitter] recognize fixupx.com URLs

mikf committed:
represent util.NONE as 'null' in JSON output

was '"None"' before

mikf committed:
[fantia] simplify 'tags' to a list of strings (#4752)

mikf committed:
[hitomi] recognize 'imageset' gallery URLs (#4756)

mikf committed:
[twitter] add test for fixupx.com

mikf committed:
merge #4755: [twitter] recognize fixupx.com URLs

mikf committed:
[exhentai] fix original image URLs (#4754)

mikf committed:
[kemonoparty] limit 'title' length (#4741)

mikf committed:
[pixiv] allow cookies for non-OAuth URLs (#4760)

mikf committed:
[nijie] set 1-2s delay between requests to avoid 429 errors

mikf committed:
[patreon] support 'id:<campaign_id>' in place of a user name

https://patreon.com/id:12345
… and remove 'campaign-id' config option

mikf committed:
reword 'reddit.client-id' instructions (#4749)

mikf committed:
[exhentai] provide fallback URLs (#1021, #4745)

mikf committed:
[exhentai] try to avoid 'DH_KEY_TOO_SMALL' errors (#1021, #4593)

mikf committed:
[pp:metadata] ignore non-string tag values (#4764)

mikf committed:
release version 1.26.2

mikf committed:
[instagram] fix exception on empty 'video_versions' (#4795)

mikf committed:
[weibo] detect redirects to login page (#4773)

mikf committed:
[twitter] ignore promoted Tweets (#4790, #3894)

add 'ads' option in case someone actually wants to
download promoted content for whatever reason

mikf committed:
[mastodon] fix reblogs (#4580)

mikf committed:
[mastodon] update test results

mikf committed:
[exhentai] handle 'Downloading … requires GP' errors (#4576, #4763)

mikf committed:
[hentaifoundry] check for and update expired sessions (#4694)

mikf committed:
[oauth] warn when cache is enabled but not writeable (#4771)

mikf committed:
rework and extend input file processing (#4732)

- add 2 command-line options to modify input file contents
  - -I/--input-file-comment
  - -x/--input-file-delete
- implement InputManager class
- move code from util.py to __init__.py
  (mainly to avoid import cycles)

mikf committed:
add '--cbz' command-line option

JSouthGB committed:
[erome] add "count" for albums

mikf committed:
reword some (internal) option text

mikf committed:
replace '--mtime-from-date' with a more generic '--mtime'

--mtime-from-date -> --mtime date
for the same effect as before

(--mtime-from-date also still works,
 but --help now lists only --mtime)

mikf committed:
replace '--ugoira-conv' etc with a general '--ugoira'

update --ugoira webm to use the same FFmpeg args as Danbooru

--ugoira-conv          -> --ugoira vp8
--ugoira-conv-lossless -> --ugoira vp9-lossless
--ugoira-conv-copy     -> --ugoira copy

(--ugoira-conv and co still work as before,
 but --help now lists only --ugoira)

mikf committed:
reorder post processing options shown by --help

mikf committed:
[tests] install yt-dlp from PyPI for Python 3.7

https://github.com/yt-dlp/yt-dlp/commit/f4b95acafcd69a50040730dfdf732e797278fdcc

mikf committed:
[downloader:http] skip files not passing filesize-min/-max (#4821)

instead of failing the download

mikf committed:
[behance] support 'text' modules (#4799)

mikf committed:
[behance] add 'modules' option (#4799)

mikf committed:
disable 'downloader.progress' when using -q/--quiet (#4810)

it didn't produce any output since output.mode is set to to "null",
but it caused some unnecessary function calls

JSouthGB committed:
[tmohentai] add support

JSouthGB committed:
[tmohentai] fix edge cases. updated archive_fmt and filename_fmt

JSouthGB committed:
[tmohentai] refactor to str.format for backwards compatibility

mikf committed:
[exhentai] fix empty api_url with '"source": "hitomi"' (#4829)

mikf committed:
support 'metadata-*' names for '*-metadata' options

For example, instead of 'url-metadata' it is now also possible to use
'metadata-url' as option name.

- metadata-url
- metadata-path
- metadata-http
- metadata-version
- metadata-parent

Nitrousoxide committed:
add simple dockerfile

Nitrousoxide committed:
Update Dockerfile

changed youtube_dl to yt-dlp and combined install line

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

Nitrousoxide committed:
Update Dockerfile

Co-authored-by: Mike Fährmann <mike_faehrmann@web.de>

mikf committed:
add 'metadata-extractor' option (#4549)

mikf committed:
[erome] tests

mikf committed:
merge #4812: [erome] add 'count' metadata field

JSouthGB committed:
[tmohentai] inherit from GalleryExtractor. refactor metadata.

JSouthGB committed:
[tmohentai] simplify url matching

mikf committed:
[blogger] inherit from BaseExtractor

- support www.micmicidol.club (#4759)

mikf committed:
[tmohentai] simplify + tests

mikf committed:
merge #4832: [tmohentai] add 'gallery' extractor (#4808)

mikf committed:
merge #4831: add simple dockerfile

mikf committed:
[tmohentai] 'categories' -> 'genres'

quite likely that the site meant 'genres' by "Genders"

enduser420 committed:
[fapello] support '.su' TLD

mikf committed:
[pixeldrain] add 'file' and 'album' extractors (#4839)

mikf committed:
[pixeldrain] add 'api-key' option (#4839)

mikf committed:
[exhentai] add 'gp' option (#4576)

mikf committed:
merge #4841: [fapello] support '.su' TLD (#4840)

mikf committed:
[docs] fix 'Example' -> 'Default' from b43be672

copy-paste mistake

mikf committed:
[tests] implement skipping/failing tests when pressing ctrl+c

mikf committed:
[sankaku] update URL patterns

mikf committed:
[warosu] fix file URLs

mikf committed:
[zerochan] fix metadata extraction

author, path, tags

mikf committed:
[xvideos] fix metadata extraction

mikf committed:
[wallpapercave] fix extraction

mikf committed:
[webtoons] fix pagination when receiving an HTTP redirect

mikf committed:
[tumblr] fix 'day' extractor

another bug caused by a383eca7

mikf committed:
[mangaread] fix extraction

mikf committed:
[imagefap] fix single image resolution

Downloading from a single image page like
https://www.imagefap.com/photo/123456789/
returned only the thumbnail URL.

mikf committed:
[hiperdex] fix 'manga' metadata

mikf committed:
[hentaicosplays] force 'https://' for download URLs

mikf committed:
[cyberdrop] update to site layout changes

mikf committed:
[foolslide] remove 'powermanga.org'

mikf committed:
[pornhub] fix 'user' metadata for gifs

mikf committed:
[tests] swap assertEqual argument order

before this, it would show test failures as
+ test value
- extracted value
when it should be the other way round

enduser420 committed:
[nitter] fix video extraction

mikf committed:
[idolcomplex] update to site layout changes

mikf committed:
update extractor test results

still not everything, but good enough for now

mikf committed:
[tests] skip test_init for BaseExtractor classes without instances

mikf committed:
merge #4855: [nitter] fix video extraction (#4853)

mikf committed:
release version 1.26.3

mikf committed:
[github] add workflow to build and push docker images

heavily inspired by and adapted from
https://github.com/danbooru/danbooru/blob/master/.github/workflows/docker-build.yaml

mikf committed:
add Python 3.12 to classifiers list

mikf committed:
[exhentai] store more cookies when logging in (#4881)

include 'igneous', 'hath_perks', etc
and not just 'ipb_member_id' and 'ipb_pass_hash' like before

JSouthGB committed:
[urlgalleries] add support

JSouthGB committed:
[urlgalleries] add support

mikf committed:
[twitter] generalize "Login Required" error (#4734, #4324)

mikf committed:
[twitter] retry API calls when their response contains errors (#4811)

mikf committed:
[nijie] fix image URLs of multi-image posts (#4876)

mikf committed:
implement '-e/--error-file' command-line option (#4732)

copying per-URL options from regular, read-only input files
does currently not work

mikf committed:
[github] fix docker workflow (#4831)

- run only on tagged commits and not on every commit to master
- use the correct github token

mikf committed:
[docker] use PAK for GHCR login (#4831)

mikf committed:
interactively prompt for passwords on login when none is provided

mikf committed:
handle 'json' parameter in Extractor.request() manually

Mainly to allow passing custom classes like util.LazyPrompt,
but also to simplify and streamline how requests handles it.

mikf committed:
fix util.dump_response to work with bytes as header values

mikf committed:
use custom HTTPBasicAuth class

to support LazyPrompt as password
and to generate the Authorization header only once
instead of for every request

mikf committed:
add 'output.errorfile' config option

mikf committed:
[urlgalleries] simplify + resolve redirects

mikf committed:
[urlgalleries] add tests

mikf committed:
merge #4886: [urlgalleries] add 'gallery' extractor (#919, #1184, #2905)

mikf committed:
[exhentai] add 'fallback-retries' option (#4792)

mikf committed:
[twitter] quick and dirty fix for /media changes (#4898)

mikf committed:
[twitter] retry API requests only for Timeout errors (#4811)

mikf committed:
implement -e/--error-file as a logging handler

similar to --write-unsupported

Tobi823 committed:
- add code for the situation when Patreon is using window.patreon = wrapInProxy({"bootstrap":' to store metadata
- refactor code to make it more readable
- output page content when the HTML structure is unknown (to make debugging easier)

Tobi823 committed:
- reformat and refactor to pass tests

Tobi823 committed:
- adapt code to current code style

Tobi823 committed:
- fix style check failure "line to long"

mikf committed:
merge #4906: [patreon] fix bootstrap data extraction (#4904)

mikf committed:
release version 1.26.4

Wiiplay123 commented:
> After I got it working, it seems lynxchan posts are all put in different folders instead of one per-thread folder. This happens with and without the postprocessor and `--write-metadata`

The extra Directory messages are updating the directory for every post in the thread.

By default, the directory is set to threadId + the text in the post, which wasn't a problem when the directory was only updated for the first post. Now it's making folders with the original thread ID + the text of the _current_ post. The metadata filename for the text post itself gets weird and ends up having the thread ID + the most recent image filename, or just "metadata" if there is none.

Wiiplay123 commented:
https://github.com/Wiiplay123/gallery-dl/commit/751eb3b4489bf00b196449678a315d134392b357
https://github.com/Wiiplay123/gallery-dl/commit/20783436f9d7399a76a3e29a2d37d1cd23feefd1

Finally got around to pushing the extra changes to my fork to make text posts on vichan work, and added a "metadata" event that is triggered on a successful metadata write.
